platform/core.windows/src/org/netbeans/core/windows/services/DialogDisplayerImpl.java

```diff
@@ -95,53 +96,46 @@ public Dialog createDialog (final DialogDescriptor d, final Frame preferredParen
         }
         return Mutex.EVENT.readAccess (new Mutex.Action<Dialog> () {
             public Dialog run () {
-                // if a modal dialog active use it as parent
-                // otherwise use the main window
-                if (NbPresenter.currentModalDialog != null) {
-                    NbDialog dlg;
-                    if (NbPresenter.currentModalDialog.isLeaf ()) {
-                        dlg = new NbDialog(d, WindowManager.getDefault ().getMainWindow ());
-                    } else {
-                        dlg = new NbDialog(d, NbPresenter.currentModalDialog);
-                    }
-                    customizeDlg(dlg);
-                    return dlg;
-                }
-                else {
-                    Window w = preferredParent;
-                    if( null == w ) {
-                        w = KeyboardFocusManager.getCurrentKeyboardFocusManager ().getActiveWindow ();
-                        if (!(w instanceof NbPresenter) || !w.isVisible()) {
-                            // undocked window is not instanceof NbPresenter although it's NetBeans's native window
-                            // all docked windows implements ModeUIBase interface
-                            if (! (w instanceof DefaultSeparateContainer.ModeUIBase)) {
-                                Container cont = SwingUtilities.getAncestorOfClass(Window.class, w);
-                                if (cont instanceof DefaultSeparateContainer.ModeUIBase) {
-                                    w = (Window) cont;
-                                } else {
-                                    // don't set non-ide window as parent
-                                    w = WindowManager.getDefault ().getMainWindow ();
-                                }
+                Window w = preferredParent;
+                if( null == w ) {
+                    w = findDialogParent();
```

Comment:
Is there a bug here?  The old code doesn't use `preferredParent` if there's a modal dialog.  Should the preferred parent be passed through to Utilities via `findDialogParent` ?

Reply:
Possibly, if there is an active modal dialog and it is not `preferredParent`. How about validating `preferredParent`?

```java
Window w = preferredParent;
if (w != null) {
    // Ensure the preferred parent is valid
    Component p = Utilities.findDialogParent(w, () -> null);
    if (p != w) {
        w = null;
    }
}
if (w == null) {
    // Existing logic from 101 onward
}
```

Reply:
Or update and call through to `findDialogParent()` in this class -

```java
private Window findDialogParent(Window preferred) {
        Component parentComponent = Utilities.findDialogParent(preferred);
    ...
```
I was wondering if we could get rid of the null check, and just do - 

```
Window w = findDialogParent(preferredParent);
```

But I think that only makes sense if we removed the checks for `NbPresenter` etc.  What are the chances we have an active modal dialog that is filtered out by that check?

Reply:
I started going down that route and `testParent` failed because it expected the input frame as the parent which it didn't get since the frame was not visible. After overriding that in the test method, it fell through to the other logic which would have it return the main window unless I customized the input frame further. Since I don't really understand the extra logic and didn't want to change the behavior I thought it would be best to just use the `preferredParent` unless it was not valid.

Reply:
@rkeen-siemens well, I've made the change as you suggested here and `testParent` still failed - seems to be another test leaving a modal dialog open.  I noticed it was passing when run by itself.

Have opened #6216 which fixes that, changes the Utilities API, and updates a couple of docs.